### PR TITLE
04 - Added play move handler

### PR DESCRIPTION
**Pull request by xavierlepretre**

### CODE REVIEW

File: x/checkers/keeper/msg_server_play_move.go [hunk 1]
```diff
@@ -2,16 +2,61 @@ package keeper
...
-	_ = ctx
+	storedGame, found := k.Keeper.GetStoredGame(ctx, msg.IdValue)
+	if !found {
+		return nil, errors.New("Game not found " + msg.IdValue)
+	}
 
-	return &types.MsgPlayMoveResponse{}, nil
+	// Is it an expected player?
+	var player rules.Player
+	if strings.Compare(storedGame.Red, msg.Creator) == 0 {
```
Comment (marbar3778) [inline x/checkers/keeper/msg_server_play_move.go:23]:
```suggestion
	if strings.Compare(storedGame.Red, msg.Creator) {
```

File: x/checkers/keeper/msg_server_play_move.go [hunk 2]
```diff
@@ -2,16 +2,61 @@ package keeper
...
+		player = rules.RED_PLAYER
+	} else if strings.Compare(storedGame.Black, msg.Creator) == 0 {
```
Comment (marbar3778) [inline x/checkers/keeper/msg_server_play_move.go:25]:
```suggestion
	} else if strings.Compare(storedGame.Black, msg.Creator) {
```

File: x/checkers/keeper/msg_server_play_move.go [hunk 3]
```diff
@@ -2,16 +2,61 @@ package keeper
...
 )
 
 func (k msgServer) PlayMove(goCtx context.Context, msg *types.MsgPlayMove) (*types.MsgPlayMoveResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	// TODO: Handling the message
-	_ = ctx
+	storedGame, found := k.Keeper.GetStoredGame(ctx, msg.IdValue)
+	if !found {
+		return nil, errors.New("Game not found " + msg.IdValue)
```
Comment (marbar3778) [inline x/checkers/keeper/msg_server_play_move.go:18]: sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"

could sdkerrors be used? 

could be something like 

```
sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "game not found %s", msg.IdValue)```

File: x/checkers/keeper/msg_server_play_move.go [hunk 4]
```diff
@@ -2,16 +2,61 @@ package keeper
...
 
-	return &types.MsgPlayMoveResponse{}, nil
+	// Is it an expected player?
+	var player rules.Player
+	if strings.Compare(storedGame.Red, msg.Creator) == 0 {
+		player = rules.RED_PLAYER
+	} else if strings.Compare(storedGame.Black, msg.Creator) == 0 {
+		player = rules.BLACK_PLAYER
+	} else {
+		return nil, errors.New("Message creator is not a player")
```
Comment (marbar3778) [inline x/checkers/keeper/msg_server_play_move.go:28]: ditto sdkerrors. 

nit: dont capitalize errors

File: x/checkers/keeper/msg_server_play_move.go [hunk 5]
```diff
@@ -2,16 +2,61 @@ package keeper
...
+	}
+
+	// Is it the player's turn?
+	fullGame := storedGame.ToFullGame()
+	if !fullGame.Game.TurnIs(player) {
+		return nil, errors.New("Player tried to play out of turn")
```
Comment (marbar3778) [inline x/checkers/keeper/msg_server_play_move.go:34]: ditto

File: x/checkers/keeper/msg_server_play_move.go (hunk 1 above)
Comment (xavierlepretre) [reply x/checkers/keeper/msg_server_play_move.go:23]: The compiler does not accept this change as this is not boolean.

File: x/checkers/keeper/msg_server_play_move.go (hunk 3 above)
Comment (xavierlepretre) [reply x/checkers/keeper/msg_server_play_move.go:18]: Among the `sdkerrors`, this `ErrUnauthorized` looks like the only one that I could use for anything related to checkers, no? Is it recommended to use these `sdkerrors`?

Comment (marbar3778) [reply x/checkers/keeper/msg_server_play_move.go:18]: yes, sdkerrors are recommended. Modules are meant to define their own type of errors as well.

Comment (xavierlepretre) [reply x/checkers/keeper/msg_server_play_move.go:18]: Nice, I see the `x/checkers/types/errors.go`. Is there a convention with the numbering?

File: x/checkers/keeper/msg_server_play_move.go (hunk 2 above)
Comment (xavierlepretre) [reply x/checkers/keeper/msg_server_play_move.go:25]: It does not compile...

File: x/checkers/types/errors.go [hunk 6]
```diff
@@ -8,8 +8,12 @@ import (
...
-	ErrInvalidCreator = sdkerrors.Register(ModuleName, 1100, "creator address is invalid")
-	ErrInvalidRed     = sdkerrors.Register(ModuleName, 1101, "red address is invalid")
-	ErrInvalidBlack   = sdkerrors.Register(ModuleName, 1102, "black address is invalid")
+	ErrInvalidCreator   = sdkerrors.Register(ModuleName, 1100, "creator address is invalid")
+	ErrInvalidRed       = sdkerrors.Register(ModuleName, 1101, "red address is invalid")
+	ErrInvalidBlack     = sdkerrors.Register(ModuleName, 1102, "black address is invalid")
+	ErrGameNotFound     = sdkerrors.Register(ModuleName, 1103, "game by id not found")
+	ErrCreatorNotPlayer = sdkerrors.Register(ModuleName, 1104, "message creator is not a player")
+	ErrNotPlayerTurn    = sdkerrors.Register(ModuleName, 1105, "player tried to play out of turn")
+	ErrWrongMove        = sdkerrors.Register(ModuleName, 1106, "wrong move")
```
Comment (xavierlepretre) [inline x/checkers/types/errors.go:17]: @marbar3778 Is this the right way to number the errors?

**Review by marbar3778 — APPROVED** [review]

**Pull request closed without merging**